internal/identity/gitlab.go

```diff
@@ -0,0 +1,127 @@
+package identity // import "github.com/pomerium/pomerium/internal/identity"
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	oidc "github.com/pomerium/go-oidc"
+	"github.com/pomerium/pomerium/internal/httputil"
+	"github.com/pomerium/pomerium/internal/sessions"
+	"github.com/pomerium/pomerium/internal/version"
+	"golang.org/x/oauth2"
+)
+
+const (
+	defaultGitLabProviderURL = "https://gitlab.com"
+	revokeURL                = "https://gitlab.com/oauth/revoke"
+	oidcURL                  = "/oauth/userinfo"
+)
+
+// GitLabProvider is an implementation of the OAuth Provider
+type GitLabProvider struct {
+	*Provider
+	Subject       string `json:"sub"`
+	Profile       string `json:"profile"`
+	Email         string `json:"email"`
+	EmailVerified bool   `json:"email_verified"`
+
+	RevokeURL string `json:"revocation_endpoint"`
+}
+
+// NewGitLabProvider returns a new GitLabProvider.
+// https://www.pomerium.io/docs/identity-providers.html#gitlab-active-directory
+func NewGitLabProvider(p *Provider) (*GitLabProvider, error) {
+	ctx := context.Background()
+
+	if p.ProviderURL == "" {
+		p.ProviderURL = defaultGitLabProviderURL
+	}
+
+	var err error
+	p.provider, err = oidc.NewProvider(ctx, p.ProviderURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(p.Scopes) == 0 {
+		p.Scopes = []string{oidc.ScopeOpenID, "api", "read_user", "profile", "email"}
+	}
+
+	p.verifier = p.provider.Verifier(&oidc.Config{ClientID: p.ClientID})
+	p.oauth = &oauth2.Config{
+		ClientID:     p.ClientID,
+		ClientSecret: p.ClientSecret,
+		Endpoint:     p.provider.Endpoint(),
+		RedirectURL:  p.RedirectURL.String(),
+		Scopes:       p.Scopes,
+	}
+	gp := &GitLabProvider{
+		Provider:  p,
+		RevokeURL: revokeURL,
+	}
+
+	if err := p.provider.Claims(&gp); err != nil {
+		return nil, err
+	}
+
+	return gp, nil
+}
+
+// Authenticate creates an identity session with gitlab from a authorization code, and makes
+// a call to the userinfo endpoint to get the information of the user.
+func (p GitLabProvider) Authenticate(ctx context.Context, code string) (*sessions.State, error) {
+	oauth2Token, err := p.oauth.Exchange(ctx, code)
+	if err != nil {
+		return nil, fmt.Errorf("internal/identity: token exchange failed: %w", err)
+	}
+
+	idToken, err := p.IdentityFromToken(ctx, oauth2Token)
+	if err != nil {
+		return nil, err
+	}
+
+	var userInfo struct {
+		ID       string   `json:"sub"`
+		Name     string   `json:"name"`
+		Nickname string   `json:"nickname"`
+		Email    string   `json:"email"`
+		Groups   []string `json:"groups"`
+	}
+
+	params := url.Values{}
+	params.Add("access_token", oauth2Token.AccessToken)
+
+	err = httputil.Client(ctx, http.MethodGet, p.ProviderURL+oidcURL, "nil", nil, params, &userInfo)
```

Comment:
This overlaps a bit with https://github.com/pomerium/pomerium/pull/529 because both require calling the user information endpoint . Would love your review on that PR, but I've confirmed this method works as well but is a little more verbose. If you are curious, check out that PR to see how the `UserInfo` endpoint can be queried from `oidc` package. 

If you are really having trouble sleeping, you can check out the RFC for the UserInfo endpoint here: https://openid.net/specs/openid-connect-core-1_0.html#UserInfo

Reply:
@desimone Using the oidc package is really less verbose compared to this current implementation, thanks.